---
title: "Life History Traits and Maximum Longevity in Mammalia"
author: "Tsz Ching Li u7630977"
date: "4 Oct 2023"
output:  
    html_document:
        toc: true
        toc_depth: 4
        theme: cosmo
        number_sections: false
        toc_float: true
        highlight: pygments
        fig_width: 8
        fig_height: 4
---
# Word/figure count

Words: [1996]
Figures: [6]

# Location on GitHub

https://github.com/sharonli126/DS4B-final-project 

# Data Description

AnAge is a database recording the longevity, ageing and life history across over 4000 animal species which researchers have collected and published relevant data and is freely available online. It was developed by researchers at the Human Ageing Genomic Resources, encompassed species data with key attributes include taxonomic classification, developmental milestone, reproductive traits , physiological characteristics, and maximum longevity. To reduce bias, the data incorporates estimated sample sizes and quality assessments. It can be widely used in biological field, such as zoology, ecology, physiology and conservation biology [1].    

# Questions/Aims

__How do life-history traits leads to diversity of longevity between four orders (Primates, Rodentia, Carivora and Artiodactyla)?__

Why do some animals live only a few years, but some can live several years, or even over a hundred year? This has been one of the most interesting questions in Biology all the time. While animals varies in lifespan, the order of Mammalia, which we, humans belong to, notably present a diverse of lifespan. For hamster or mouse, they only have around 4 years lifespan but for a gorilla, it can live up to around 60 years, or humans and whales can live even over 100 and 200 years old! 

There are various of life-history traits in the class of mammalia, such as Primates tend to have longer lifespans, extended juvenile dependency periods, and fewer offspring, while most rodents reproduce quickly, have shorter lifespans, and produce more offspring in their lifetime. Meanwhile, species of Carnivora, like lions and bears, exhibit a range of lifespans and reproductive strategies, but have the lifespan is relatively higher than Rodentia. Similarly, Artiodactyla, which deer and cattle belong to, also presents its own unique life-history patterns. [2,3]

Therefore, to understand deeper of which trait correlate the most with lifespan in different order of mammalia, this study analyse a big database of aging and life history traits in over 4600 animal species from AnAge, aiming to examine and compare the different life history traits impact on maximum longevity among mammalia, particularly in four larger order group, including Primates, Rodentia, Carnivora and Artiodactyla.

# Raw data

```{r, include = F}
library(readr)
library(skimr)
library(janitor)
library(tidyverse)
library(knitr)
```

Visit the AnAge database at https://genomics.senescence.info/species/dataset.zip. Download the dataset.zip file. Extract the anage_data.txt file from the downloaded zip archive.

```{r}
raw_data <- read_delim("raw_data/anage_data.txt", 
                   delim = "\t", 
                   escape_double = FALSE, 
                   trim_ws = TRUE)

```



# Data wrangling

Since the data is originally "each variable forms a column, each observation forms a row and each type of observational unit forms a table", so only column names are tidied.
```{r}
tidied_data <- raw_data |>
  clean_names()

write.csv(tidied_data, 
          "/Users/lisharon/Documents/GitHub/DB4S_final_project/processed_data/tidied_data.csv", 
          row.names = FALSE)

```

Considering the reliability of this study, only trait with more than 0.55 complete rate will be examined and order with more than 150 rows will be observed, for filtering an acceptable sample size.

```{r}
mammalia <- tidied_data[tidied_data$class == "Mammalia",]

mammalia |> group_by(order) |> summarise(n = n()) |> arrange(desc(n))
```

```{r}
skim(mammalia) |> filter(complete_rate > 0.55)
```

Therefore, the order of Rodentia, Carnivora, Primates and Artiodactyla are selected and trait including day of female maturity, gestation and weaning, number of babies per time and year, and adult weight.

```{r}
# Filter data need for this report

mammalia_subset <- mammalia[, c("hagrid", "kingdom", "phylum", "class", "order", "family", "genus", "species", "female_maturity_days", "gestation_incubation_days", "weaning_days", "litter_clutch_size", "litters_clutches_per_year", "adult_weight_g", "maximum_longevity_yrs" )]

combined_order <- mammalia_subset[mammalia_subset$order %in% c("Rodentia", "Carnivora", "Primates", "Artiodactyla"),]
```

Some weird datas were identified and noted in the excluded data, through grouping all value by genus, mean and range of max and min value were found. Suspected genus with a range close to mean will searched online. All data vary from information online will be noted in the excluded data.

```{r}
excluded_data <- read_csv("processed_data/excluded_data.csv")
```

NA is remained, since a larger sample size is better for EDA and correlation testing.

```{r}
cleaned_data <- combined_order |> 
  filter(!hagrid %in% excluded_data$hagrid)

write.csv(cleaned_data, 
          "/Users/lisharon/Documents/GitHub/DB4S_final_project/processed_data/cleaned_data.csv", 
          row.names = FALSE)
```

# Sanity checks

The missing value and the complete rate should still be similar as before.

```{r}
skim(cleaned_data)
```

All string are identified and checked to be match.

```{r}
str(cleaned_data)
```


# Addressing the questions/aims

## Exploratory Data Analysis

### Histogram of Maximum Longevity of the Four Orders
```{r}
ggplot(long_cleaned_data, aes(x = maximum_longevity_yrs)) +
  geom_histogram() +
  labs(x = "Maximum Longevity (year)") +
  scale_x_continuous(breaks=seq(0, 120, by=20)) +
  facet_wrap(~ order) +
  theme_minimal()
```

According to the histogram, the maximum longevity of Primates is spread mainly between 10 to 60 years, with few at above 60 or over 120 years. Carnivora has a similar distribution as Artiodactyiam whereas both of them have a range from 10 to 60, but Artiodactyia has a larger count in from 20-40.  While longevity of Rodentia are all below 45 years, and half of them are below 10 years of lifespan, which is obviously lower than all the others in general. 

In this study, one of the challenge faced is the presence of missing value in data. To address this, both missing value retained data and removed data are anaylsed by visuliation, which allow pattern in both data be observed and compared.

Both cleaned_data and cleaned_na_rm_data are pivoted longer for plotting.

```{r}
long_cleaned_data <- cleaned_data |> 
  pivot_longer(c(female_maturity_days, gestation_incubation_days, 
                 weaning_days, litter_clutch_size, 
                 litters_clutches_per_year, adult_weight_g), 
               names_to = "trait",
               values_to = "value") |>
  mutate(trait = recode(trait, # code in mutate functioni from chatGPT
                        female_maturity_days = "Female Maturity (days)",
                        gestation_incubation_days = "Pregnancy Period (days)",
                        weaning_days = "Stopped Milk Feeding (days)", 
                        litter_clutch_size = "Number of Babies in each Pregnancy",
                        litters_clutches_per_year = "Number of Babies per Year", 
                        adult_weight_g = "Adult Weight (g)"))
```

```{r}
long_cleaned_na_rm_data <- cleaned_na_rm_data |> 
  pivot_longer(c(female_maturity_days, gestation_incubation_days, 
                 weaning_days, litter_clutch_size, 
                 litters_clutches_per_year, adult_weight_g), 
               names_to = "trait",
               values_to = "value") |>
  mutate(trait = recode(trait, 
                        female_maturity_days = "Female Maturity (days)",
                        gestation_incubation_days = "Pregnancy Period (days)",
                        weaning_days = "Stopped Milk Feeding (days)", 
                        litter_clutch_size = "Number of Babies in each Pregnancy",
                        litters_clutches_per_year = "Number of Babies per Year", 
                        adult_weight_g = "Adult Weight (g)")) 
```

### Correlation of Different Traits and Maximum Longevity (with missing value)

```{r, fig.height=10}
ggplot(long_cleaned_data, aes(x = value, y = maximum_longevity_yrs, color = order)) +
  geom_point(aes(color = order), alpha = .1) +
  geom_smooth(aes(color = order), method = "lm", alpha = .5) +
  labs(y = "Max. Longevity (year", x = "Value of each Trait") +
  facet_wrap(~ trait, scales = "free", ncol = 2) +
  theme_minimal()
```

### Correlation of Different Traits and Maximum Longevity (without missing value)

```{r, fig.height=10}
ggplot(long_cleaned_na_rm_data, aes(x = value, y = maximum_longevity_yrs)) +
  geom_point(aes(color = order), alpha = .1) +
  geom_smooth(aes(color = order), method = "lm", alpha = .5) +
  labs(y = "Max. Longevity (year)", x = "Value of each Trait") +
  facet_wrap(~ trait, scales = "free", ncol = 2) +
  theme_minimal()
```

As compared with the plotting with missing value, the trend of all trait does not change, but most of its slope changed due to less sample size has shifted the correlation. 

## Correlation Test

The correlation test (CT) can provide a quick look on the strength and direction of linear relationship among each trait and lifespan. Using the data with missing value, a more accurate relationship and trend can be identified in its larger sample size.

```{r}
get_cor <- function(data, trait) {
  
  cor.test(data$maximum_longevity_yrs, data[[trait]])$estimate
  
}

data = cleaned_data
  
overall_cor <- c(get_cor(data, "female_maturity_days"), 
                 get_cor(data, "gestation_incubation_days"), 
                 get_cor(data, "weaning_days"),
                 get_cor(data, "litter_clutch_size"),
                 get_cor(data, "litters_clutches_per_year"),
                 get_cor(data, "adult_weight_g"))

summary_cor <- data.frame(Trait = c("Female Maturity (day)", "Pregnancy Period (day)", "Milk Feeding Stop (day)", 
                                    "Number of Babies per Pregnancy", "Number of Babies per Year", "Adult Weight (g)"),
                          Overall = round(overall_cor, 3)
                          )

for (i in unique(long_cleaned_data$order)) {
  
  filtered <- cleaned_data[cleaned_data$order == i,]
    
  current_cor <- c(get_cor(filtered, "female_maturity_days"), 
                 get_cor(filtered, "gestation_incubation_days"), 
                 get_cor(filtered, "weaning_days"), 
                 get_cor(filtered, "litter_clutch_size"), 
                 get_cor(filtered, "litters_clutches_per_year"), 
                 get_cor(filtered, "adult_weight_g"))
  
  summary_cor[[i]] <- round(current_cor, 3)
  
}

kable(summary_cor)

```

The CT have reflected the same trend as the two correlation plot in EDA section. All order have the same direction of correlation in each trait with female maturity and pregnancy period as the highest correlated in overall. However, different degree of correlation were found in individual order testing, such as adult weight is having the highest correlation in the order of Artiodactyla but it is not the highest correlated trait in other orders. 
 
## Linear model

Linear model (LM) can statistically show which independent variables (life history trait/ physiological characteristic) significantly associate with dependent variable (maximum longevity) for mammalia in general and each order.

To fit a LM, there should be no missing value, so that the trait significance can be compared more accurately.

```{r}
na_rm_data <- cleaned_data |>
  na.omit()
```

```{r}
library(broom)

linear_model_pvalue <- function(data) {
  model <- lm(data = data,
            maximum_longevity_yrs ~ 
              female_maturity_days +
              weaning_days +
              gestation_incubation_days +
              litter_clutch_size +
              litters_clutches_per_year +
              adult_weight_g
              ) |> 
    tidy()
  
  pvalue <- model$p.value[-1]
  
  return(pvalue)
}

summary_lm <- data_frame(Trait = c("Female Maturity (day)", "Pregnancy Period (day)", "Milk Feeding Stop (day)", 
                                    "Number of Babies per Pregnancy", "Number of Babies per Year", "Adult Weight (g)"),
                         Overall = linear_model_pvalue(na_rm_data)
)

for (i in unique(na_rm_data$order)) {
  
  filtered <- na_rm_data[na_rm_data$order == i,]
    
  
  summary_lm[[i]] <- linear_model_pvalue(filtered)
  
}

kable(summary_lm)
```

```{r}
convert_to_stars <- function(p) {
  ifelse(p < 0.001, "***",
         ifelse(p < 0.01, "**",
                ifelse(p < 0.05, "*",
                       ifelse(p < 0.1, ".", " "))))
}

summary_lm_word <- summary_lm %>%
  mutate_at(vars(-Trait), convert_to_stars)

kable(summary_lm_word, caption = "Significance of Linear Model Fitting")

```

Note:

"***": Highly Significant

"**" : Very Significant 

"*"  : Significant

"."  : Suggestion of Association

" "  : Not Significant

The LM have found female maturity, number of babies each pregnancy and per year, and adult weight are significant associated with the maximum longevity overall (among all the 4 species) with the female maturity as the most significant. This is probably due to its reflection on the growth rate of a species, whereas species take longer time to mature usually live longer with a lower growth rate [5]. 

Meanwhile, when fitting data from four orders separately into the LM, distinct associations emerge for each trait across the orders. The female maturity only remain the most significant in Carnivora, while gestation period stand out among Primates and Rodentia, and adult weight for Artiodactyla. 

Pregnancy period in determining the lifespan of Primates and Rodentia can be understood through its representation of parental investment Many mammal, especially the larger ones usually adopt "K-strategies" in reproduction. This refers to reproducing slower and producing less offspring with greater care and investment in each offspring, contrast with the "R-strategies" [3,6]. Moreover, for Artiodactyla, they are usually a prey in the wild, larger size of them are generally less prone to predation and leading to a lower mortality rate, hence, extending their lifespan but this is not an universal trend in mammals [5].

Comparing the result from LM and CT, they are not alligned, such as adult weight is highly significant but is having the least positive correlation overall, and both Primates and Rodentia are having highest degree of correlation with weaning day (milk feeding stop) but it is not significant in linear model. This result might due to the different, since the LM is using the data without missing value, which amount is less than a half of data with missing value. 

However, in conclusion, all visulisation, CT, and LM have shown the relationship of different traits with maximum longevity and these correlations are varied in different orders, the most significant trait associated with their lifespan can be different too. With AnAge becoming more complete and larger, I believe a more accurate and wide linking between different life-history traits and lifespan can be analysed.

# References

1. A database of vertebrate longevity records and their relation to other life-history traits. https://doi.org/10.1111/j.1420-9101.2009.01783.x

2. Taxonomic differences in the mammalian life span-body weight relationship and the problem of brain weight. doi: 10.1159/000212400.

3. Evolution of Reproductive Life History in Mammals and the Associated Change of Functional Constraints.  https://doi.org/10.3390/genes12050740

4. Human Ageing Genomic Resources: integrated databases and tools for the biology and genetics of ageing. https://doi.org/10.1093/nar/gks1155

5. An analysis of the relationship between metabolism, developmental schedules, and longevity using phylogenetic independent contrasts. https://doi.org/10.1093/gerona/62.2.149

6. Parental investment, late reproduction, and increased reserve capacity are associated with longevity in humans. doi: 10.2114/jpa2.25.119. 
