---
title: "Life History Traits and Maximum Longevity in Mammalia"
author: "Tsz Ching Li u7630977"
date: "4 Oct 2023"
output:  
    html_document:
        toc: true
        toc_depth: 4
        theme: cosmo
        number_sections: false
        toc_float: true
        highlight: pygments
        fig_width: 8
        fig_height: 4
---
# Word/figure count

Words: [1999]
Figures: [6]

# Location on GitHub

https://github.com/sharonli126/DS4B-final-project 

# Data Description

AnAge is a database recording the longevity, ageing and life history across over 4600 animal species which researchers have collected and published relevant data and is freely available online. It was developed by researchers at the Human Ageing Genomic Resources, encompassed species data with key attributes include taxonomic classification, developmental milestone, reproductive traits , physiological characteristics, and maximum longevity. To minimise bias, the data incorporates estimated sample sizes and quality assessments. It can be widely used in biological field, such as zoology, ecology, and conservation biology [1].    

# Questions/Aims

__Which life-history traits are most significantly associated with variations in maximum longevity among the orders of Primates, Rodentia, Carnivora, and Artiodactyla?__

Why do some animals live only few years, but some can live several years, or even over a hundred year? This has been one of the most interesting questions in Biology all the time. While animals varies in lifespan, the order of Mammalia, which we, humans belong to, notably present a diversity of lifespan. For mouse, they only have around 4 years lifespan but for a gorilla, it can live up to around 60 years, or humans and whales can live even over 100 and 200 years old! 

There are various of life-history traits in the class of mammalia, such as Primates tend to have longer lifespans, extended juvenile dependency periods, and fewer offspring, while most rodents is in contrast, and have more offspring in their lifetime. Meanwhile, species of Carnivora and Artiodactyla, like lions and deer respectively, exhibit a range of lifespans and reproductive strategies, but their lifespan is relatively higher than Rodentia [2,3].

Therefore, to understand which trait correlate the most with lifespan in different order of mammalia, this study analyse the big database of aging and life-history traits from AnAge, aiming to examine and compare life-history traits impact on maximum longevity among mammalia, particularly in four larger order group, including Primates, Rodentia, Carnivora and Artiodactyla.

# Raw data

```{r, include = F}
library(readr)
library(skimr)
library(janitor)
library(tidyverse)
library(knitr)
```

Visit the AnAge database at https://genomics.senescence.info/species/dataset.zip. Download the dataset.zip file. Extract the anage_data.txt file from the downloaded zip archive.

```{r}
raw_data <- read_delim("raw_data/anage_data.txt", 
                   delim = "\t", 
                   escape_double = FALSE, 
                   trim_ws = TRUE)

```


# Data wrangling

Since the data is originally "each variable = column, each observation = row and each type of observational unit forms a table", so only column names are tidied.
```{r}
tidied_data <- raw_data |>
  clean_names()

write.csv(tidied_data, 
          "/Users/lisharon/Documents/GitHub/DB4S_final_project/processed_data/tidied_data.csv", 
          row.names = FALSE)

```

Considering the reliability of this study, only trait with more than 0.55 complete rate will be examined and order with more than 150 rows will be observed for filtering an acceptable sample size.

```{r}
mammalia <- tidied_data[tidied_data$class == "Mammalia",]

mammalia |> group_by(order) |> summarise(n = n()) |> arrange(desc(n))
```

```{r}
skim(mammalia) |> filter(complete_rate > 0.55)
```

Therefore, the top four orders are filtered and trait including day of female maturity, gestation and weaning, number of babies per time and year, and adult weight.

```{r}
# Filter data for this report

mammalia_subset <- mammalia[, c("hagrid", "kingdom", "phylum", "class", "order", "family", "genus", "species", "female_maturity_days", "gestation_incubation_days", "weaning_days", "litter_clutch_size", "litters_clutches_per_year", "adult_weight_g", "maximum_longevity_yrs" )]

combined_order <- mammalia_subset[mammalia_subset$order %in% c("Rodentia", "Carnivora", "Primates", "Artiodactyla"),]
```

Suppose species from same genus group will have a similar trait, therefore, some weird observations were identified and noted in the excluded data, through grouping all value by genus, summeraising mean, range of max-min value were found. 

For example, a suspected genus, Ateles in Primates has a range of max-min weaning_day (411), which is half of the mean (675) will be spotted and searched online, due to its variation within the genus group.

```{r}
excluded_data <- read_csv("processed_data/excluded_data.csv")
```

NA is remained, since a larger sample size is better for EDA and correlation testing.

```{r}
cleaned_data <- combined_order |> 
  filter(!hagrid %in% excluded_data$hagrid)

write.csv(cleaned_data, 
          "/Users/lisharon/Documents/GitHub/DB4S_final_project/processed_data/cleaned_data.csv", 
          row.names = FALSE)
```

# Sanity checks

The missing value and the complete rate should still be similar as before.

```{r}
skim(cleaned_data)
```

All string are identified and checked to be match.

```{r}
str(cleaned_data)
```


# Addressing the questions/aims

## Exploratory Data Analysis

In this study, one of the challenge faced is the presence of missing value in data. To address this, both missing value retained data and removed data are analysed by visualization, which allow pattern in both data be observed and compared.

Both cleaned_data and cleaned_na_rm_data are pivoted longer for plotting.

```{r}
long_mutate <- function(data){
  data|>
    pivot_longer(c(female_maturity_days, gestation_incubation_days, 
                 weaning_days, litter_clutch_size, 
                 litters_clutches_per_year, adult_weight_g), 
               names_to = "trait",
               values_to = "value") |>
  mutate(trait = recode(trait, # code in mutate function from chatGPT
                        female_maturity_days = "Female Maturity (days)",
                        gestation_incubation_days = "Pregnancy Period (days)",
                        weaning_days = "Stopped Milk Feeding (days)", 
                        litter_clutch_size = "Number of Babies in each Pregnancy",
                        litters_clutches_per_year = "Number of Babies per Year", 
                        adult_weight_g = "Adult Weight (g)"))
}

long_cleaned_data <- long_mutate(cleaned_data)
```

```{r}
na_rm_data <- cleaned_data |>
  na.omit()
```

```{r}
long_na_rm_data <- long_mutate(na_rm_data)
```

### Histogram of Maximum Longevity of the Four Orders

```{r}
ggplot(long_cleaned_data, aes(x = maximum_longevity_yrs)) +
  geom_histogram() +
  labs(x = "Maximum Longevity (year)") +
  scale_x_continuous(breaks=seq(0, 120, by=20)) +
  facet_wrap(~ order) +
  theme_minimal()
```

According to the histogram, the maximum longevity of Primates is spread mainly between 10 to 60 years, with few at above 60 or over 120 years. Carnivora has a similar distribution as Artiodactyiam whereas both of them have a range from 10 to 60, but Artiodactyia has a larger count in from 20-40.  While longevity of Rodentia are all below 45 years, and half of them are below 10 years of lifespan, which is obviously lower than all the others in general. 


### Correlation of Different Traits and Maximum Longevity (with missing-value)

```{r, fig.height=10}
ggplot(long_cleaned_data, aes(x = value, y = maximum_longevity_yrs, color = order)) +
  geom_point(alpha = .1) +
  geom_smooth(method = "lm", alpha = .3) +
  labs(y = "Maximum Longevity (year)", x = "Value of Trait") +
  facet_wrap(~ trait, scales = "free", ncol = 2) +
  theme_minimal()
```

The visualization above have shown the linear relationship of each traits with maximum longevity. All directions are the same in each plot, but all traits have exhibit a different slope in all plots, except the female maturity, which all have a similar slope, indicating this trait has a similar impact on the lifespan in all analysed orders. 

### Correlation of Different Traits and Maximum Longevity (without missing-value)

```{r, fig.height=10}
ggplot(long_na_rm_data, aes(x = value, y = maximum_longevity_yrs, color = order)) +
  geom_point(alpha = .1) +
  geom_smooth(method = "lm", alpha = .5) +
  labs(y = "Maximum Longevity (year)", x = "Value of Trait") +
  facet_wrap(~ trait, scales = "free", ncol = 2) +
  theme_minimal()
```

As compared with the plotting with missing value, the trend of most trait did not change, only the slope of Artioactyle (red) in the number of babies in each pregnancy has shift to a positive correlation, but with a non-steep slope. While other slope of trait are probably changed due to less sample size has shifted the correlation, such as Rodentia (purple) in day of milk feeding stop plot. Therefore, the following test and fitting are expected to be varied, with application of two different cleaned data.

## Correlation Test

The correlation test (CT) can provide a quick look on the strength and direction of linear relationship between traits and lifespan. Using the data with missing value, a more accurate relationship and trend can be identified in its larger sample size.

```{r}
get_cor <- function(data, trait) {
  
  # function to obtain "estimation in cor.test
  cor.test(data$maximum_longevity_yrs, data[[trait]])$estimate
  
}

trait = c("Female Maturity (day)", "Pregnancy Period (day)", "Milk Feeding Stop (day)", 
          "Number of Babies per Pregnancy", "Number of Babies per Year", "Adult Weight (g)")

data = cleaned_data
  
overall_cor <- c(get_cor(data, "female_maturity_days"), 
                 get_cor(data, "gestation_incubation_days"), 
                 get_cor(data, "weaning_days"),
                 get_cor(data, "litter_clutch_size"),
                 get_cor(data, "litters_clutches_per_year"),
                 get_cor(data, "adult_weight_g"))

summary_cor <- data.frame(Trait = trait,
                          Overall = round(overall_cor, 3)
                          )

for (i in unique(long_cleaned_data$order)) {
  
  filtered <- cleaned_data[cleaned_data$order == i,]
    
  current_cor <- c(get_cor(filtered, "female_maturity_days"), 
                 get_cor(filtered, "gestation_incubation_days"), 
                 get_cor(filtered, "weaning_days"), 
                 get_cor(filtered, "litter_clutch_size"), 
                 get_cor(filtered, "litters_clutches_per_year"), 
                 get_cor(filtered, "adult_weight_g"))
  
  summary_cor[[i]] <- round(current_cor, 3)
  
}

kable(summary_cor)

```

The CT have reflected a similar trend as the first correlation plots in EDA section. All order have the same direction of correlation in each trait with female maturity and pregnancy period as the highest correlated in overall. However, different degree of correlation were found in individual order testing, such as adult weight is having the highest correlation in the order of Artiodactyla but it is not the highest correlated trait in other orders. 
 
## Linear model

Linear model (LM) can statistically show which independent variables (life history trait/ physiological characteristic) significantly associate with dependent variable (maximum longevity) for mammalia in general and each order.

To fit a LM, there should be no missing value, so that the trait significance can be compared more accurately.

```{r}
library(broom)

linear_model_pvalue <- function(data) {
  model <- lm(data = data,
            maximum_longevity_yrs ~ 
              female_maturity_days +
              weaning_days +
              gestation_incubation_days +
              litter_clutch_size +
              litters_clutches_per_year +
              adult_weight_g
              ) |> 
    tidy()
  
  pvalue <- model$p.value[-1]
  
  return(pvalue)
}

summary_lm <- data_frame(Trait = trait,
                         Overall = linear_model_pvalue(na_rm_data)
)

for (i in unique(na_rm_data$order)) {
  
  filtered <- na_rm_data[na_rm_data$order == i,]
    
  
  summary_lm[[i]] <- linear_model_pvalue(filtered)
  
}

# I would like to look at the exact value, so kable is not used here
summary_lm
```

```{r}
# the function covert value to star, so that the significane of correlation can easier to visualise

convert_to_stars <- function(p) {
  ifelse(p < 0.001, "***",
         ifelse(p < 0.01, "**",
                ifelse(p < 0.05, "*",
                       ifelse(p < 0.1, ".", " "))))
}

summary_lm_word <- summary_lm %>%
  mutate_at(vars(-Trait), convert_to_stars)

kable(summary_lm_word, caption = "Significance of Linear Model Fitting")

```

Note:

"***": Highly Significant

"**" : Very Significant 

"*"  : Significant

"."  : Suggestion of Association

" "  : Not Significant

The LM have found female maturity, number of babies each pregnancy and per year, and adult weight are highly significant associated with the maximum longevity overall (among all the 4 species) with the female maturity as the most significant. This is probably due to its reflection on the growth rate of a species, whereas species take longer time to mature usually live longer with a lower growth rate [5]. 

Meanwhile, when fitting data from four orders separately into the LM, distinct associations emerge for each trait across the orders. The female maturity only remain the most significant in Carnivora, which probably due to its position often as a predator, can afford the trade off to live longer with a delayed maturity, increasing the fitness of individual [7]. 

While the stand out of pregnancy (gestation) period in determining the lifespan of Primates and Rodentia can be understood through its representation of parental investment. Many mammal, especially the larger ones usually adopt "K-strategies" in reproduction. This refers to reproducing slower and producing less offspring with greater care and investment in each offspring, contrast with the "R-strategies" [3,6]. Species with a higher parental investment are more likely to live longer, this is particularly stand out in Primates and Rodentia [3]. 

For the significance of adult weight in Artiodactyla, they are usually a prey in the wild compared to other order, larger size of them are generally less prone to predation and leading to a lower mortality rate, hence, extending their lifespan [5].

Comparing the result from LM and CT, they are not aligned, such as adult weight is highly significant but is having the least positive correlation overall, and both Primates and Rodentia are having the highest degree of correlation with weaning day (milk feeding stop) but it is not significant in linear model. This result might due to the different data used, where as the observation of na_rm_data is a half less than the cleaned_data 

## Conclusion

Through various analysis, this study highlight the significance of certain life-history traits, with female maturity day standing out as a primary determinant of longevity across the order. However, with AnAge becoming more complete and larger, I believe a more accurate and wide linking between different life-history traits and lifespan can be analysed.

# References

A_database_of_vertebrate_longevity_records_and_their_relation_to_other_life-history_traits. https://doi.org/10.1111/j.1420-9101.2009.01783.x

Taxonomic_differences_in_the_mammalian_life_span-body_weight_relationship_and_the_problem_of_brain_weight.doi: 10.1159/000212400.

Evolution_of_Reproductive_Life_History_in_Mammals_and_the_Associated_Change_of_Functional_Constraints.  https://doi.org/10.3390/genes12050740

Human_Ageing_Genomic_Resources:_integrated_databases_and_tools_for_the_biology_and_genetics_of_ageing. https://doi.org/10.1093/nar/gks1155

An_analysis_of_the_relationship_between_metabolism,_developmental_schedules,_and_longevity_using_phylogenetic_independent_contrasts. https://doi.org/10.1093/gerona/62.2.149

Parental_investment,_late_reproduction,_and_increased_reserve_capacity_are_associated_with_longevity_in_humans.doi: 10.2114/jpa2.25.119. 

The_fast–slow_continuum_and_mammalian_life-history_patterns:_an_empirical_evaluation. https://doi.org/10.1016/j.baae.2004.06.002

chatGPT
